core/src/main/java/software/amazon/awssdk/SdkRequest.java

```diff
@@ -15,12 +15,63 @@
 
 package software.amazon.awssdk;
 
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+import java.util.Optional;
+
 /**
  * The base class for all SDK requests.
  *
- * TODO: SDK-specific options on the {@link AmazonWebServiceRequest} should be migrated here as part of the base-model refactor.
- *
  * @see SdkResponse
  */
-public abstract class SdkRequest {
+public abstract class SdkRequest<B extends SdkRequest.Builder<B, R, RequestOverrideConfigT>,
+        R extends SdkRequest<B, R, RequestOverrideConfigT>,
```

Comment:
Do we ever think we'll need to convert a non-concrete request back to a builder and then back to the request type? It would be nice if we didn't need these parameters.

We could also just remove the parent interfaces, define the methods as an abstract `Builder toBuilder()` on the request and `SdkRequest build()` on the builder and let the child classes implement the interfaces?

I figure we'll be using SdkRequest a lot, and SdkRequest<?, ?> is pretty cumbersome. We pretty much never care about the parameters. The only impact would be that we couldn't use the default methods that are defined on the interface for the SdkRequest.

Reply:
> We could also just remove the parent interfaces,

You mean `CopyableBuilder`/`ToCopyableBuilder`? I'd be okay with that if there's no objections. I just wanted to be consistent with the rest of the codebase since we use them pretty extensively.

Reply:
I agree, the loss of consistency is annoying. It's definitely just a personal preference thing. If people would prefer the consistency, I'm cool with that.

I'm a bit leery of having generics in places that we use the base class a lot (in this case, more than the child classes), especially if we are pretty much only ever going to specify the parameters as wildcards.

Reply:
> Do we ever think we'll need to convert a non-concrete request back to a builder and then back to the request type?

As a maintainer of a Scala library that uses [generic programming to abstract pagination](https://github.com/Dwolla/scala-aws-utils/blob/fs2/fs2AwsUtils/src/main/scala/com/dwolla/awssdk/PaginatedAwsClient.scala) (creating [fs2 `Stream`](https://github.com/functional-streams-for-scala/fs2) for a given list of data (e.g. CloudFormation stacks or ECS services)), I would say yes, it is helpful to be able to convert a generic request back to a builder of the appropriate type in order to create a new request with modifications.

With the 1.11 SDK, I'm using structural types to identify what [`AmazonWebServiceRequest`](https://github.com/Dwolla/scala-aws-utils/blob/fs2/core/src/main/scala/com/dwolla/awssdk/utils/package.scala#L12) and [`AmazonWebServiceResponse`](https://github.com/Dwolla/scala-aws-utils/blob/fs2/core/src/main/scala/com/dwolla/awssdk/utils/package.scala#L14) types are paginated. Given a base request of the right type and a reference to an [async request function](https://github.com/Dwolla/scala-aws-utils/blob/fs2/core/src/main/scala/com/dwolla/awssdk/utils/package.scala#L15), it can fetch all the pages for the caller and provide the data as an fs2 stream. This relies on being able to construct a new request object for each page it needs to request, in order to set the pagination token for the next request.

Right now I'm hacking around it by passing around a request factory function `() ⇒ PaginatedRequest`, where 

```scala
type PaginatedRequest = AmazonWebServiceRequest {def setNextToken(s: String): Unit}
```

(that's a nullary function that returns an instance of type `PaginatedRequest`, which is a subtype of `AmazonWebServiceRequest` with method `setNextToken` having the right signature), which is called to create each request, and then modified with the next pagination token.

I'm still trying to wrap my head around the structure of the new types, but I think it would definitely be useful to be able to create new instances of a given type rather than have to pass around the factory function I'm using now.

I'm not sure what the right level of detail is here. I would just ask that as you make design decisions of this type, you keep this type of thing in mind. Let me know how I can be more helpful.

(FWIW, it would also be nice to have an interface for pagination so I don't have to rely on structural typing (which uses reflection at runtime, with the corresponding runtime penalties) to deal with the pagination tokens.)

Reply:
That's great info, thanks! I think it definitely makes sense that people should be able to do this then, at least at the `AwsRequest` level, which is the level most people care about.

> FWIW, it would also be nice to have an interface for pagination so I don't have to rely on structural typing (which uses reflection at runtime, with the corresponding runtime penalties) to deal with the pagination tokens.

Because you've given great feedback, I feel like I want to ask for more. Would you mind giving yours on our automatic depagination pull request to see if it fits your needs? https://github.com/aws/aws-sdk-java-v2/pull/207